### ISSUE TITLE
Infer const-ness of local let bindings that aren't assigned to

### ISSUE THREAD

**Issue by marijnh** [microsoft/TypeScript#39251]
## Search Terms

let const infer assignment

## Suggestion

In some cases, type narrowing for `const` bindings is more effective than for `let` bindings. Specifically, using such bindings from a function that closes over them will keep the narrowed type, which doesn't happen for `let` binding, presumably on the assumption that those might be mutated at any time.

This suggests to treat local `let` bindings that are never assigned to (which can be determined by a rather straightforward static check on the syntax tree) like `const` bindings for this purpose.

(Interestingly, something like this seems to already exist for function parameters.)

## Use Cases

Though some people have adopted a coding style where locals are declared with `const` unless assigned to, there's also a good argument to be made that that is not a very good use of brain cycles and just using `let` for all locals is fine. This proposal would improve type inference for people using the `let` style.

## Examples

```
function foo() {
  let v: number[] | null = Math.random() < 0.5 ? [1] : null
  if (Array.isArray(v)) setTimeout(() => console.log(v.length), 100)
}
```

The compiler complains that `v` is possibly null. If the binding is changed to `const` the problem goes away. If `v` is made a function parameter it also doesn't occur. (But comes back when a `v = null` statement is added, suggesting a check like the one suggested here is being done in that case.)

## Checklist

My suggestion meets these guidelines:

* [x] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, etc.)
* [x] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

**Comment by RyanCavanaugh** [comment]
This would be a breaking change, but I can see the appeal

**Comment by marijnh** [comment]
Is accepting more programs a breaking change? Or am I missing some case where this would reject something it currently accepts?

**Comment by RyanCavanaugh** [comment]
You'd induce a second error in this program, for example:
```ts
function f1() {
    let s = Math.random() > 0.5 ?  "hello" : "goodbye";
    return s;
}
let a = ([f1()] as const)[0];
a = "world";

function f2() {
    const s = Math.random() > 0.5 ?  "hello" : "goodbye";
    return s;
}
let b = ([f2()] as const)[0];
b = "world";
```

**Comment by marijnh** [comment]
Ah, I see. But at a glance it seems like the type inference for initial binding is a different operation from narrowing of that type down the line, and possibly this could be executed in a backwards-compatible way by only doing the latter?

**Comment by ahejlsberg** [comment]
I agree, this wouldn't necessarily be a breaking change because we'd still widen types when inferring the declared type of a `let` variable. My concerns are more around performance impact. It's not a cheap check. Also, previous requests for this feature generally asked for analysis that no assignments occur following creation of the function closure, as opposed to no assignments at all other than the initialization of the variable. That is a much harder problem to solve, and not one that we could solve with guaranteed correctness.

**Comment by DanielRosenwasser** [comment]
To provide more context - in our compiler, declarations don't keep track of use-sites, so knowing whether something is even written to isn't even information we have on-hand. If we wanted to get it eagerly, there are a couple of hacks I could imagine, but we believe it might increase memory pressure (at least, that's what @mhegazy told me was the reason for why we didn't have all use-sites available for find-all-references).

If we wanted to get them lazily, there's still some concern with how much work is done on-demand. We have a goal in our design that it should be possible to resolve and know the type of a given variable at an arbitrary position in a way that doesn't require checking type-checking across the whole world (or even the whole subtree of a declaration). If figuring out the type of an identifier required looking up all the use-sites, that might mean we'd have to go to the declaration, get its container, and do a full walk of the container to look for every identifier with the same name.

**Comment by marijnh** [comment]
If there is a scope resolution phase or similar, where identifiers are associated with declarations, and you have a single bit free in the flags for declaration items, I believe this could be implemented by allocating an is-assigned-to bit in the declaration and setting that when coming across an assignment. This is always file-local, for local variables. But I may be missing something.

**Comment by DanielRosenwasser** [comment]
I'd considered the same thing, but dismissed it because I was thinking of tracking all the use-sites - but you're right, global files aside, you only need to track the first encounter of a write in the binder (the symbol table setup phase). Not every variable has been declared in scope yet, but for anything unresolved we could keep it around until the end of the scope. I'm not sure how much overhead that would be, but it might be worth experimenting with.

Another concern I'm starting to realize is just the UX. First, there's the idea that writes have non-local effects. The idea that a later write could change the type earlier than the write itself feels like it might be kind of surprising to me.

Then there's just consistency with object literals: the idea that

```ts
let foo = "hello"
```

would get the type `"hello"` seems reasonable, but then a user might expect the same of `foo.prop`

let foo = {
  prop: "hello"
};
```

and we can't make the same sort of will-get-written-to checks based on the above strategy.

**Comment by marijnh** [comment]
Makes sense. I'd argue that getting type changes when adding assignment is not much different from getting type changes when switching between `let` and `const`, but I can imagine this being annoying when such type changes leak into some external interface. This is definitely not the no-brainer I initially assumed it to be, so feel free to close at your discretion.
